README.md
A script to generate SNMPv3 keys as detailed by rfc3414 (passphrases expanded with a kdf, then hashed with the engine id). All key generation is done using the new cryptographically secure _secrets_ library.

As I wrote this with ESXi in mind, it also emits a string suitable for configuring the SNMP daemon via esxcli/PowerCLI, but the hashes are standard and compatible with other SNMPv3 implementations.

With no arguments, it will generate an authentication and privacy passphrase with associated random engine ID in text form. `--json` will format the output as json.

The script is fully idempotent; if you take the parameters it generates randomly and re-enter them, you will get the same output a second time.

Dependencies
============
Runtime: Python 3.6 or greater.

Sample Output
=============

Standard
--------

```
User: observium
Auth: gaYA82XVtNaf3WLwRgoIs544ghP6f80S / f78359764ca382922fa382cf884e588031de575a
Priv: H5XEtRpxXVaGzXU5i2rFwPnYGr8SEzTp / 31a001a56a225fdfc1916bd60190405a1aa22ff0
Engine: 7ae1b0ff0aa2f3950566d3de2274d05a
ESXi USM String: observium/f78359764ca382922fa382cf884e588031de575a/31a001a56a225fdfc1916bd60190405a1aa22ff0/authpriv
```

JSON
----

```
{
  "user": "observium",
  "engine": "b2a50167b7c8512ddfc9d5765a3490af",
  "phrases": {
    "auth": "71rOhjfj6QVSy2mw5tBo7PueZ8KWSv60",
    "priv": "xwsvzht8NEcuwAlEpUKzMxKFWeH72sK9"
  },
  "hashes": {
    "auth": "fa0d5249293404502f9953b9514d0636a96c2cbc",
    "priv": "cccbdcfa603817df340514ecc22dfae8c4c412e8"
  },
  "esxi": "observium/fa0d5249293404502f9953b9514d0636a96c2cbc/cccbdcfa603817df340514ecc22dfae8c4c412e8/authpriv"}
```

If a YAML or TOML library is installed, you also use the `--yaml` and `--toml` arguments respectively.

It should go without saying, but **DO NOT** use the engine id or passphrases in the samples.

Usage
=====

```
usage: snmpv3-hashgen [-h] [--auth AUTH] [--priv PRIV] [--engine ENGINE] [--user USER] [--mode {auth,priv,none}] [--hash {md5,sha1,sha224,sha256,sha384,sha512}] [--json | --yaml | --toml]

Convert an SNMPv3 auth or priv passphrase to hashes.

optional arguments:
  -h, --help            show this help message and exit
  --auth AUTH           Authentication passphrase to be derived as utf8 string
  --priv PRIV           Privacy passphrase to be derived as utf8 string
  --engine ENGINE       Engine ID as hex string
  --user USER           SNMPv3 USM username (default "librenms")
  --mode {auth,priv,none}
                        SNMPv3 mode (default "priv")
  --hash {md5,sha1,sha224,sha256,sha384,sha512}
                        Hash algorithm to use (default "sha1")
  --json                Emit output as json
  --yaml                Emit output as yaml
  --toml                Emit output as toml

RFC 7630 defines no test data for sha[2-9]{3} - these should be considered experimental.
Report bugs at https://github.com/TheMysteriousX/SNMPv3-Hash-Generator/issues
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.1] - 2020-07-12

This release has no functional changes

### Changed

- Changed casing of YAML
- Changed how the runtime version is detected

## [0.4.0] - 2020-07-12

### Added

- Added a changelog
- Added some automated testing using Github actions

### Changed

- Switched to Semver - 0.4.0 didn't change the API, just the versioning scheme.

## [0.3] - 2020-07-12

### Added

- YAML and TOML included as output options if the relevant libraries are installed
- Untested support for RFC 7630 hashes (SHA-2 224, 256, 384 and 512)

### Changed

- Switched the default username to 'librenms'

### Fixed

- Method binding issue preventing use of md5
- Kdf was hard coded to sha1

## [0.2] - 2017-04-13

### Fixed

- Confusion between authPriv, authpriv, auth, priv and none


tests/tests.py
import unittest

from snmpv3_hashgen.hashgen import Hashgen

# flake8: noqa: E501


class TestLibrary(unittest.TestCase):
    kdf_maplesyrups = {
        "md5": "9faf3283884e92834ebc9847d8edd963",  # RFC 3413 §A.3.1
        "sha1": "9fb5cc0381497b3793528939ff788d5d79145211",  # RFC 3413 §A.3.2
        "sha224": "282a5867ee9aac639ad59df9572c7d3ac0fbc13a905b6df07dbbf00b",  #  Unverified
        "sha256": "ab51014d1e077f6017df2b12bee5f5aa72993177e9bb569c4dff5a4ca0b4afac",  # Unverified
        "sha384": "e06eccdf2c68a06ed034723c9c26e0db3b669e1e2efed49150b55377a2e98f383c86fb836857444654b287c93f51ff64",  # Unverified
        "sha512": "7e4396de5aadc77be853819b98c9406265b3a9c37cc3176569847a4e4f6fba63dd3a73d04924d31a63f95a601f9385af6be4ed1b37f87d040f7c6ed6f8d38a91",  # Unverified
    }

    final_maplesyrups = {
        "md5": "526f5eed9fcce26f8964c2930787d82b",  # RFC 3413 §A.3.1
        "sha1": "6695febc9288e36282235fc7151f128497b38f3f",  # RFC 3413 §A.3.2
        "sha224": "0bd8827c6e29f8065e08e09237f177e410f69b90e1782be682075674",  # Unverified
        "sha256": "8982e0e549e866db361a6b625d84cccc11162d453ee8ce3a6445c2d6776f0f8b",  # Unverified
        "sha384": "3b298f16164a11184279d5432bf169e2d2a48307de02b3d3f7e2b4f36eb6f0455a53689a3937eea07319a633d2ccba78",  # Unverified
        "sha512": "22a5a36cedfcc085807a128d7bc6c2382167ad6c0dbc5fdff856740f3d84c099ad1ea87a8db096714d9788bd544047c9021e4229ce27e4c0a69250adfcffbb0b",  # Unverified
    }

    test_string = "maplesyrup"
    engine_id = "000000000000000000000002"

    def test_kdf(self):
        for alg, fptr in Hashgen.algs.items():
            self.assertEqual(Hashgen.kdf(self.test_string, fptr).hex(), self.kdf_maplesyrups[alg])

    def test_concat(self):
        for alg, fptr in Hashgen.algs.items():
            self.assertEqual(Hashgen.derive_msg(self.test_string, self.engine_id, fptr).hex(), "".join([self.kdf_maplesyrups[alg], self.engine_id, self.kdf_maplesyrups[alg]]))

    def test_final(self):
        for alg, fptr in Hashgen.algs.items():
            self.assertEqual(fptr(Hashgen.derive_msg(self.test_string, self.engine_id, fptr)), self.final_maplesyrups[alg])


snmpv3_hashgen/hashgen.py
import hashlib
import string
import secrets

from itertools import repeat
from functools import partial

P_LEN = 32
E_LEN = 16


class Hashgen(object):
    @staticmethod
    def hash(bytes, alg=hashlib.sha1, raw=False):
        digest = alg(bytes).digest()
        return digest if raw else digest.hex()

    @staticmethod
    def expand(substr, target_len):
        reps = target_len // len(substr) + 1  # approximation; worst case: overrun = l + len(s)
        return "".join(list(repeat(substr, reps)))[:target_len]

    @staticmethod
    def kdf(password, alg=None):
        alg = Hashgen.algs["sha1"] if alg is None else alg

        data = Hashgen.expand(password, 1048576).encode("utf-8")
        return alg(data, raw=True)

    @staticmethod
    def random_string(len=P_LEN, alphabet=(string.ascii_letters + string.digits)):
        return "".join(secrets.choice(alphabet) for _ in range(len))

    @staticmethod
    def random_engine(len=E_LEN):
        return secrets.token_hex(len)

    @staticmethod
    def derive_msg(passphrase, engine, alg):
        # Parameter derivation á la rfc3414
        Ku = Hashgen.kdf(passphrase, alg)
        E = bytearray.fromhex(engine)

        return b"".join([Ku, E, Ku])


# Define available hash algorithms
Hashgen.algs = {
    "md5": partial(Hashgen.hash, alg=hashlib.md5),
    "sha1": partial(Hashgen.hash, alg=hashlib.sha1),
    "sha224": partial(Hashgen.hash, alg=hashlib.sha224),
    "sha256": partial(Hashgen.hash, alg=hashlib.sha256),
    "sha384": partial(Hashgen.hash, alg=hashlib.sha384),
    "sha512": partial(Hashgen.hash, alg=hashlib.sha512),
}


scripts/snmpv3_hashgen.py
#! /usr/bin/env python3

import sys

if sys.version_info < (3, 6):
    print("Python 3.6 or higher is required, please see https://www.python.org/ or your OS package repository", file=sys.stderr)
    sys.exit(2)

import argparse
import json

from snmpv3_hashgen import Hashgen

help_text = """
Convert an SNMPv3 auth or priv passphrase to hashes.
"""

epilog_text = """
RFC 7630 defines no test data for sha[2-9]{3} - these should be considered experimental.
Report bugs at https://github.com/TheMysteriousX/SNMPv3-Hash-Generator/issues
"""

parser = argparse.ArgumentParser(formatter_class=argparse.RawDescriptionHelpFormatter, description=help_text, epilog=epilog_text,)
parser.add_argument("--auth", type=str, help="Authentication passphrase to be derived as utf8 string")
parser.add_argument("--priv", type=str, help="Privacy passphrase to be derived as utf8 string")
parser.add_argument("--engine", type=str, help="Engine ID as hex string")
parser.add_argument("--user", type=str, help='SNMPv3 USM username (default "librenms")')
parser.add_argument("--mode", type=str, choices=["auth", "priv", "none"], help='SNMPv3 mode (default "priv")')
parser.add_argument("--hash", type=str, choices=["md5", "sha1", "sha224", "sha256", "sha384", "sha512"], help='Hash algorithm to use (default "sha1")')

fmt = parser.add_mutually_exclusive_group()
fmt.add_argument("--json", action="store_true", help="Emit output as json")
fmt.add_argument("--yaml", action="store_true", help="Emit output as yaml")
fmt.add_argument("--toml", action="store_true", help="Emit output as toml")


def format_esxi(user, Kul_auth, Kul_priv, mode, hash):
    if mode == "priv":
        return f"{user}/{hash(Kul_auth)}/{hash(Kul_priv)}/{mode}"
    elif mode == "auth":
        return f"{user}/{hash(Kul_auth)}/-/{mode}"
    else:
        return f"{user}/-/-/{mode}"


def main(*args, **kwargs):
    # Argument setup
    args = parser.parse_args()

    if args.priv and not args.auth:
        print("Error: privacy passphrase supplied without auth passphrase", file=sys.stderr)
        sys.exit(3)

    user = "librenms" if not args.user else args.user
    mode = "priv" if not args.mode else args.mode
    auth = Hashgen.random_string() if not args.auth else args.auth
    priv = Hashgen.random_string() if not args.priv else args.priv
    engine = Hashgen.random_engine() if not args.engine else args.engine
    hash = Hashgen.algs["sha1"] if not args.hash else Hashgen.algs[args.hash]

    #  Derive Kul from passphrases
    try:
        Kul_auth = Hashgen.derive_msg(auth, engine, hash) if "none" not in mode else None
        Kul_priv = Hashgen.derive_msg(priv, engine, hash) if "priv" in mode else None
    except ValueError:
        print("Error: Engine ID seems invalid; ensure that it is entered as a hex character string", file=sys.stderr)
        sys.exit(1)

    esxi = format_esxi(user, Kul_auth, Kul_priv, mode, hash)
    output = {
        "user": user,
        "engine": engine,
        "phrases": {"auth": auth if "none" not in mode else None, "priv": priv if "priv" in mode else None},
        "hashes": {"auth": hash(Kul_auth) if "auth" in mode or "priv" in mode else None, "priv": hash(Kul_priv) if "priv" in mode else None},
        "esxi": esxi,
    }

    if args.json:
        print(json.dumps(output))
    elif args.yaml:
        try:
            import yaml

            print(yaml.dump({"snmpv3": output}, explicit_start=True))
        except ImportError:
            print("YAML output requires a YAML library")
            print("Try running pip3 install PyYAML")
            sys.exit(4)
    elif args.toml:
        try:
            import toml

            print(toml.dumps(output))
        except ImportError:
            print("TOML output requires a TOML library")
            print("Try running pip3 install toml")
            sys.exit(5)
    else:
        print(f"User: {user}")
        if "none" not in mode:
            print(f"Auth: {auth} / {hash(Kul_auth)}")
        if "priv" in mode:
            print(f"Priv: {priv} / {hash(Kul_priv)}")
        print(f"Engine: {engine}")
        print(f"ESXi USM String: {esxi}")


if __name__ == "__main__":
    main()
